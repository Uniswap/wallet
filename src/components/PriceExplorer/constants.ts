import { HistoryDuration } from 'src/data/__generated__/types-and-hooks'
import { dimensions } from 'src/styles/sizing'
import { theme as FixedTheme } from 'src/styles/theme'

// sets the height of the chart short enough on small devices that the
// "Your balance" section will always show above the fold
// we can't use useResponsiveProps for this because CHART_HEIGHT gets
// used in non-component code related to chart functionality
export const CHART_HEIGHT = dimensions.fullHeight < FixedTheme.breakpoints.sm.height ? 180 : 310
export const CHART_WIDTH = dimensions.fullWidth

export const NUM_GRAPHS = 5

export const BUTTON_PADDING = 20
export const BUTTON_WIDTH = CHART_WIDTH / NUM_GRAPHS
export const LABEL_WIDTH = BUTTON_WIDTH - BUTTON_PADDING * 2

export const CURSOR_INNER_SIZE = 12
export const CURSOR_SIZE = CURSOR_INNER_SIZE + 6
export const LINE_WIDTH = 1

export const TIME_RANGES = [
  [HistoryDuration.Hour, '1H'],
  [HistoryDuration.Day, '1D'],
  [HistoryDuration.Week, '1W'],
  [HistoryDuration.Month, '1M'],
  [HistoryDuration.Year, '1Y'],
] as const
