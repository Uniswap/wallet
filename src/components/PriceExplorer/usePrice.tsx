import { SharedValue, useDerivedValue } from 'react-native-reanimated'
import {
  useLineChart,
  useLineChartPrice as useRNWagmiChartLineChartPrice,
} from 'react-native-wagmi-charts'
import { numberToLocaleStringWorklet, numberToPercentWorklet } from 'src/utils/reanimated'

export type ValueAndFormatted<U = number, V = string> = {
  value: Readonly<SharedValue<U>>
  formatted: Readonly<SharedValue<V>>
}

/**
 * Wrapper around react-native-wagmi-chart#useLineChartPrice that returns price
 * when not scrubbing.
 */
export function useLineChartPrice({ spotPrice }: { spotPrice?: number }): ValueAndFormatted {
  // `price` when scrubbing the chart
  const { value: activeCursorPrice } = useRNWagmiChartLineChartPrice()

  const price = useDerivedValue(() => {
    if (activeCursorPrice.value) {
      return Number(activeCursorPrice.value)
    }

    return spotPrice ?? 0
  })

  const priceFormatted = useDerivedValue(() => {
    return numberToLocaleStringWorklet(price.value, 'en-US', {
      style: 'currency',
      currency: 'USD',
    })
  })

  return {
    value: price,
    formatted: priceFormatted,
  }
}

export function useLineChartRelativeChange({
  spotRelativeChange,
}: {
  spotRelativeChange?: number
}): ValueAndFormatted {
  const { currentIndex, data, isActive } = useLineChart()

  const relativeChange = useDerivedValue(() => {
    // when not scrubbing, return relative change 24h
    if (!isActive.value) {
      return spotRelativeChange ?? 0
    }

    // when scrubbing, compute relative change from close price
    const activePrice = data[currentIndex.value]?.value
    const closePrice = data[data.length - 1]?.value

    if (!activePrice || !closePrice) {
      return 0
    }

    return ((activePrice - closePrice) / closePrice) * 100
  })

  const relativeChangeFormattted = useDerivedValue(() => {
    return numberToPercentWorklet(relativeChange.value, { precision: 2, absolute: true })
  })

  return { value: relativeChange, formatted: relativeChangeFormattted }
}
