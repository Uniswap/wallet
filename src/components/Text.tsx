import { createText, useResponsiveProp } from '@shopify/restyle'
import React, { ComponentProps, PropsWithChildren } from 'react'
import { useWindowDimensions } from 'react-native'
import Animated from 'react-native-reanimated'
import { Box } from 'src/components/layout'
import { Shimmer } from 'src/components/loading/Shimmer'
import { HiddenFromScreenReaders } from 'src/components/text/HiddenFromScreenReaders'
import { textVariants } from 'src/styles/font'
import { Theme } from 'src/styles/theme'

export const DEFAULT_FONT_SCALE = 1

export type TextProps = ComponentProps<typeof ThemedText> & {
  maxFontSizeMultiplier?: number
  animated?: boolean
  allowFontScaling?: boolean
  loading?: boolean | 'no-shimmer'
  loadingPlaceholderText?: string
}

// Use this text component throughout the app instead of
// Default RN Text for theme support
const ThemedText = createText<Theme>()
const ThemedAnimatedText = createText<Theme>(Animated.Text)

const TextPlaceholder = ({ children }: PropsWithChildren<unknown>): JSX.Element => {
  return (
    <Box alignItems="center" flexDirection="row">
      <Box alignItems="center" flexDirection="row" position="relative">
        <HiddenFromScreenReaders>{children}</HiddenFromScreenReaders>
        <Box
          bg="background3"
          borderRadius="rounded4"
          bottom="5%"
          left={0}
          position="absolute"
          right={0}
          top="5%"
        />
      </Box>
    </Box>
  )
}

const TextLoaderWrapper = ({
  children,
  loadingShimmer,
}: { loadingShimmer?: boolean } & PropsWithChildren<unknown>): JSX.Element => {
  const inner = <TextPlaceholder>{children}</TextPlaceholder>
  if (loadingShimmer) {
    return <Shimmer>{inner}</Shimmer>
  }

  return inner
}

/**
 * Use this component instead of the default React Native <Text> component anywhere text shows up throughout the app, so we can use the design system values for colors and sizes, and make sure all text looks and behaves the same way
 * @param loading Whether the text inside the component is still loading or not. Set this to true if whatever content goes inside the <Text> component is coming from a variable that might still be loading. This prop is optional and defaults to false. This prop can also be set to "no-shimmer" to enable a loading state without the shimmer effect.
 * @param loadingPlaceholderText - The text that the loader's size will be derived from. Pick something that's close to the same length as the final text is expected to be, e.g. if it's a ticker symbol, "XXX" might be a good placeholder text. This prop is optional and defaults to "$00.00".
 */
export const Text = ({
  animated,
  loading = false,
  maxFontSizeMultiplier,
  allowFontScaling,
  loadingPlaceholderText = '$00.00',
  ...rest
}: TextProps): JSX.Element => {
  const { fontScale } = useWindowDimensions()
  const variant = useResponsiveProp(rest.variant ?? 'bodySmall') as keyof typeof textVariants
  const enableFontScaling = allowFontScaling ?? fontScale > DEFAULT_FONT_SCALE
  const multiplier = maxFontSizeMultiplier ?? textVariants[variant].maxFontSizeMultiplier

  if (animated) {
    return (
      <ThemedAnimatedText
        allowFontScaling={enableFontScaling}
        maxFontSizeMultiplier={multiplier}
        {...rest}
      />
    )
  }

  if (loading) {
    return (
      <TextLoaderWrapper loadingShimmer={loading !== 'no-shimmer'}>
        <ThemedText
          allowFontScaling={enableFontScaling}
          color="none"
          maxFontSizeMultiplier={multiplier}
          opacity={0}
          {...rest}>
          {/* Important that `children` isn't used or rendered by <Text> when `loading` is true, because if the child of a <Text> component is a dynamic variable that might not be finished fetching yet, it'll result in an error until it's finished loading. We use `loadingPlaceholderText` to set the size of the loading element instead. */}
          {loadingPlaceholderText}
        </ThemedText>
      </TextLoaderWrapper>
    )
  }

  return (
    <ThemedText allowFontScaling={enableFontScaling} maxFontSizeMultiplier={multiplier} {...rest} />
  )
}
