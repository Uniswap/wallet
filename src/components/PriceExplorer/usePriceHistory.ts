import { Dispatch, SetStateAction, useCallback, useMemo, useState } from 'react'
import { TLineChartData } from 'react-native-wagmi-charts'
import { PollingInterval } from 'src/constants/misc'
import { isError, isNonPollingRequestInFlight } from 'src/data/utils'
import {
  HistoryDuration,
  TimestampedAmount,
  useTokenPriceHistoryQuery,
} from 'src/data/__generated__/types-and-hooks'
import { GqlResult } from 'src/features/dataApi/types'
import { currencyIdToContractInput } from 'src/features/dataApi/utils'

/**
 * @returns Token price history for requested duration
 */
export function useTokenPriceHistory(
  currencyId: string,
  initialDuration: HistoryDuration = HistoryDuration.Day
): Omit<
  GqlResult<{ priceHistory?: TLineChartData; spot: { value?: number; relativeChange?: number } }>,
  'error'
> & {
  setDuration: Dispatch<SetStateAction<HistoryDuration>>
  error: boolean
} {
  const [duration, setDuration] = useState(initialDuration)

  const {
    data: priceData,
    refetch,
    networkStatus,
  } = useTokenPriceHistoryQuery({
    variables: {
      contract: currencyIdToContractInput(currencyId),
      duration,
    },
    notifyOnNetworkStatusChange: true,
    pollInterval: PollingInterval.Normal,
    fetchPolicy: 'cache-first',
  })

  const retry = useCallback(() => {
    refetch({ contract: currencyIdToContractInput(currencyId) })
  }, [refetch, currencyId])

  const { price, pricePercentChange24h, priceHistory } =
    priceData?.tokenProjects?.[0]?.markets?.[0] ?? {}
  const formattedPriceHistory = useMemo(
    () =>
      priceHistory
        ?.filter((x): x is TimestampedAmount => Boolean(x))
        .map((x) => ({ timestamp: x.timestamp * 1000, value: x.value })),
    [priceHistory]
  )
  const spot = useMemo(
    () => ({
      value: price?.value,
      relativeChange: pricePercentChange24h?.value,
    }),
    [price?.value, pricePercentChange24h?.value]
  )

  return useMemo(
    () => ({
      data: { priceHistory: formattedPriceHistory, spot },
      loading: isNonPollingRequestInFlight(networkStatus),
      error: isError(networkStatus, !!priceData),
      refetch: retry,
      setDuration,
    }),
    [formattedPriceHistory, networkStatus, priceData, retry, spot]
  )
}
