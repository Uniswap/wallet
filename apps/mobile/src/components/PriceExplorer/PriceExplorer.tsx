import { ImpactFeedbackStyle } from 'expo-haptics'
import React, { useMemo } from 'react'
import { SharedValue } from 'react-native-reanimated'
import {
  LineChart,
  LineChartProvider,
  TLineChartData,
  TLineChartDataProp,
} from 'react-native-wagmi-charts'
import { Loader } from 'src/components/loading'
import { CURSOR_INNER_SIZE, CURSOR_SIZE } from 'src/components/PriceExplorer/constants'
import { PriceExplorerError } from 'src/components/PriceExplorer/PriceExplorerError'
import { DatetimeText, PriceText, RelativeChangeText } from 'src/components/PriceExplorer/Text'
import { TimeRangeGroup } from 'src/components/PriceExplorer/TimeRangeGroup'
import { useChartDimensions } from 'src/components/PriceExplorer/useChartDimensions'
import { invokeImpact } from 'src/utils/haptic'
import { Flex } from 'ui/src'
import { HistoryDuration } from 'wallet/src/data/__generated__/types-and-hooks'
import { useFiatConverter } from 'wallet/src/features/fiatCurrency/conversion'
import { CurrencyId } from 'wallet/src/utils/currencyId'
import { TokenSpotData, useTokenPriceHistory } from './usePriceHistory'

type PriceTextProps = {
  loading: boolean
  relativeChange?: SharedValue<number>
}

function PriceTextSection({ loading, relativeChange }: PriceTextProps): JSX.Element {
  return (
    <Flex mx="$spacing12">
      <PriceText loading={loading} />
      <Flex row gap="$spacing4">
        <RelativeChangeText loading={loading} spotRelativeChange={relativeChange} />
        <DatetimeText loading={loading} />
      </Flex>
    </Flex>
  )
}

export type LineChartPriceAndDateTimeTextProps = {
  currencyId: CurrencyId
}

export function PriceExplorer({
  currencyId,
  tokenColor,
  forcePlaceholder,
  onRetry,
}: {
  currencyId: string
  tokenColor?: string
  forcePlaceholder?: boolean
  onRetry: () => void
}): JSX.Element {
  const { data, loading, error, refetch, setDuration, selectedDuration } =
    useTokenPriceHistory(currencyId)
  const { convertFiatAmount } = useFiatConverter()
  const conversionRate = convertFiatAmount().amount
  const convertedPriceHistory = useMemo(
    (): TLineChartData | undefined =>
      data?.priceHistory?.map((point) => {
        return { ...point, value: point.value * conversionRate }
      }),
    [data, conversionRate]
  )
  const convertedSpot = useMemo((): TokenSpotData | undefined => {
    return (
      data?.spot && {
        ...data?.spot,
        value: { value: conversionRate * (data?.spot?.value?.value ?? 0) },
      }
    )
  }, [data, conversionRate])

  if (
    !loading &&
    (!convertedPriceHistory || (!convertedSpot && selectedDuration === HistoryDuration.Day))
  ) {
    // Propagate retry up while refetching, if available
    const refetchAndRetry = (): void => {
      if (refetch) refetch()
      onRetry()
    }
    return <PriceExplorerError showRetry={error !== undefined} onRetry={refetchAndRetry} />
  }

  let content: JSX.Element | null
  if (forcePlaceholder) {
    content = <PriceExplorerPlaceholder loading={forcePlaceholder} />
  } else if (convertedPriceHistory?.length) {
    content = (
      <PriceExplorerChart
        loading={loading}
        priceHistory={convertedPriceHistory}
        spot={convertedSpot}
        tokenColor={tokenColor}
      />
    )
  } else {
    content = <PriceExplorerPlaceholder loading={loading} />
  }

  return (
    <Flex overflow="hidden">
      {content}
      <TimeRangeGroup setDuration={setDuration} />
    </Flex>
  )
}

function PriceExplorerPlaceholder({ loading }: { loading: boolean }): JSX.Element {
  return (
    <Flex gap="$spacing8">
      <PriceTextSection loading={loading} />
      <Flex my="$spacing24">
        <Loader.Graph />
      </Flex>
    </Flex>
  )
}

function PriceExplorerChart({
  priceHistory,
  spot,
  loading,
  tokenColor,
}: {
  priceHistory: TLineChartDataProp
  spot?: TokenSpotData
  loading: boolean
  tokenColor?: string
}): JSX.Element {
  const { chartHeight, chartWidth } = useChartDimensions()

  return (
    <LineChartProvider
      data={priceHistory}
      onCurrentIndexChange={invokeImpact[ImpactFeedbackStyle.Light]}>
      <Flex gap="$spacing8">
        <PriceTextSection loading={loading} relativeChange={spot?.relativeChange} />
        <Flex my="$spacing24">
          <LineChart height={chartHeight} width={chartWidth}>
            <LineChart.Path color={tokenColor} pathProps={{ isTransitionEnabled: false }} />
            <LineChart.CursorLine color={tokenColor} />
            <LineChart.CursorCrosshair
              color={tokenColor}
              outerSize={CURSOR_SIZE}
              size={CURSOR_INNER_SIZE}
              onActivated={invokeImpact[ImpactFeedbackStyle.Light]}
              onEnded={invokeImpact[ImpactFeedbackStyle.Light]}
            />
          </LineChart>
        </Flex>
      </Flex>
    </LineChartProvider>
  )
}
