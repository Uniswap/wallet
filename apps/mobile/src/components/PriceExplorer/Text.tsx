import React from 'react'
import { SharedValue, useAnimatedStyle, useDerivedValue } from 'react-native-reanimated'
import { useLineChartDatetime } from 'react-native-wagmi-charts'
import { useAppTheme } from 'src/app/hooks'
import { Flex } from 'src/components/layout'
import { Text } from 'src/components/Text'
import { AnimatedText } from 'src/components/text/AnimatedText'
import { AnimatedDecimalNumber } from './AnimatedDecimalNumber'
import { useLineChartPrice, useLineChartRelativeChange } from './usePrice'

export function PriceText({ loading }: { loading: boolean }): JSX.Element {
  const price = useLineChartPrice()

  if (loading) {
    return <Text loading loadingPlaceholderText="$10,000" variant="headlineLarge" />
  }

  return <AnimatedDecimalNumber number={price} testID="price-text" variant="headlineLarge" />
}

export function RelativeChangeText({
  loading,
  spotRelativeChange,
}: {
  loading: boolean
  spotRelativeChange?: SharedValue<number>
}): JSX.Element {
  const theme = useAppTheme()

  const relativeChange = useLineChartRelativeChange({ spotRelativeChange })
  const icon = useDerivedValue(() => (relativeChange.value.value > 0 ? '↗' : '↘'))
  const styles = useAnimatedStyle(() => ({
    color:
      relativeChange.value.value > 0 ? theme.colors.accentSuccess : theme.colors.accentCritical,
  }))

  if (loading) {
    return <Text loading loadingPlaceholderText="00.00%" variant="bodyLarge" />
  }

  return (
    <Flex row gap="spacing2">
      <AnimatedText style={styles} testID="relative-change-icon" text={icon} variant="bodyLarge" />
      <AnimatedText
        style={styles}
        testID="relative-change-text"
        text={relativeChange.formatted}
        variant="bodyLarge"
      />
    </Flex>
  )
}

export function DatetimeText({ loading }: { loading: boolean }): JSX.Element | null {
  // `datetime` when scrubbing the chart
  const datetime = useLineChartDatetime()

  if (loading) return null

  return <AnimatedText color="textSecondary" text={datetime.formatted} variant="bodyLarge" />
}
