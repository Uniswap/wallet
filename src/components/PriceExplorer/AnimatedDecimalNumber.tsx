import React, { useMemo } from 'react'
import { useAnimatedStyle, useDerivedValue } from 'react-native-reanimated'
import { useAppTheme } from 'src/app/hooks'
import { TextProps } from 'src/components/Text'
import { AnimatedText } from 'src/components/text/AnimatedText'
import { Theme } from 'src/styles/theme'
import { Box } from '../layout'
import { ValueAndFormatted } from './usePrice'

type AnimatedDecimalNumberProps = TextProps & {
  number: ValueAndFormatted
  separator?: string
  variant: keyof Theme['textVariants']
  wholePartColor?: string
  decimalPartColor?: string
  decimalThreshold?: number // below this value (not including) decimal part would have wholePartColor too
}

// Utility component to display decimal numbers where the decimal part
// is dimmed using AnimatedText
export function AnimatedDecimalNumber(props: AnimatedDecimalNumberProps): JSX.Element {
  const theme = useAppTheme()

  const {
    number,
    separator = '.',
    variant,
    wholePartColor = theme.colors.textPrimary,
    decimalPartColor = theme.colors.textTertiary,
    decimalThreshold = 1,
    ...rest
  } = props

  const wholePart = useDerivedValue(
    () => number.formatted.value.split(separator)[0] || '',
    [number, separator]
  )
  const decimalPart = useDerivedValue(
    () => separator + (number.formatted.value.split(separator)[1] || ''),
    [number, separator]
  )

  const wholeStyle = useMemo(() => {
    return {
      color: wholePartColor,
    }
  }, [wholePartColor])

  const decimalStyle = useAnimatedStyle(() => {
    return {
      color: number.value.value < decimalThreshold ? wholePartColor : decimalPartColor,
    }
  }, [decimalThreshold, wholePartColor, decimalPartColor])

  return (
    <Box flexDirection="row" {...rest}>
      <AnimatedText style={wholeStyle} testID="wholePart" text={wholePart} variant={variant} />
      <AnimatedText
        style={decimalStyle}
        testID="decimalPart"
        text={decimalPart}
        variant={variant}
      />
    </Box>
  )
}
