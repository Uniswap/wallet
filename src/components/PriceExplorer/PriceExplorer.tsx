import { ImpactFeedbackStyle } from 'expo-haptics'
import React from 'react'
import { SharedValue } from 'react-native-reanimated'
import { LineChart, LineChartProvider } from 'react-native-wagmi-charts'
import { Flex } from 'src/components/layout'
import { Box } from 'src/components/layout/Box'
import { Loader } from 'src/components/loading'
import {
  CHART_HEIGHT,
  CURSOR_INNER_SIZE,
  CURSOR_SIZE,
} from 'src/components/PriceExplorer/constants'
import { PriceExplorerError } from 'src/components/PriceExplorer/PriceExplorerError'
import { DatetimeText, PriceText, RelativeChangeText } from 'src/components/PriceExplorer/Text'
import { TimeRangeGroup } from 'src/components/PriceExplorer/TimeRangeGroup'
import { CurrencyId } from 'src/utils/currencyId'
import { invokeImpact } from 'src/utils/haptic'
import { useTokenPriceHistory } from './usePriceHistory'

type PriceTextProps = {
  loading: boolean
  relativeChange?: SharedValue<number>
}

function PriceTextSection({ loading, relativeChange }: PriceTextProps): JSX.Element {
  return (
    <Box mx="spacing12">
      <PriceText loading={loading} />
      <Flex row gap="spacing4">
        <RelativeChangeText loading={loading} spotRelativeChange={relativeChange} />
        <DatetimeText loading={loading} />
      </Flex>
    </Box>
  )
}

export type LineChartPriceAndDateTimeTextProps = {
  currencyId: CurrencyId
}

export function PriceExplorer({
  currencyId,
  tokenColor,
  onRetry,
}: {
  currencyId: string
  tokenColor?: string
  onRetry: () => void
}): JSX.Element {
  const { data, loading, error, refetch, setDuration } = useTokenPriceHistory(currencyId)

  if (!loading && !data) {
    // Propagate retry up while refetching, if available
    const refetchAndRetry = (): void => {
      if (refetch) refetch()
      onRetry()
    }
    return <PriceExplorerError showRetry={error !== undefined} onRetry={refetchAndRetry} />
  }

  return (
    <Box overflow="hidden">
      {data?.priceHistory ? (
        <LineChartProvider
          data={data.priceHistory}
          onCurrentIndexChange={invokeImpact[ImpactFeedbackStyle.Light]}>
          <Flex gap="spacing8">
            <PriceTextSection loading={loading} relativeChange={data.spot?.relativeChange} />
            <Box my="spacing24">
              <LineChart height={CHART_HEIGHT}>
                <LineChart.Path color={tokenColor} />

                <LineChart.CursorLine color={tokenColor} />
                <LineChart.CursorCrosshair
                  color={tokenColor}
                  outerSize={CURSOR_SIZE}
                  size={CURSOR_INNER_SIZE}
                  onActivated={invokeImpact[ImpactFeedbackStyle.Light]}
                  onEnded={invokeImpact[ImpactFeedbackStyle.Light]}
                />
              </LineChart>
            </Box>
          </Flex>
        </LineChartProvider>
      ) : (
        <Flex gap="spacing8">
          <PriceTextSection loading={loading} />
          <Box my="spacing24">
            <Loader.Graph />
          </Box>
        </Flex>
      )}

      <TimeRangeGroup setDuration={setDuration} />
    </Box>
  )
}
