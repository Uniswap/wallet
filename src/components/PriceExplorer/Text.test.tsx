import React from 'react'
import * as charts from 'react-native-wagmi-charts'
import { DatetimeText, PriceText, RelativeChangeText } from 'src/components/PriceExplorer/Text'
import { Amounts } from 'src/test/gqlFixtures'
import { render } from 'src/test/test-utils'

jest.mock(
  'react-native-wagmi-charts'
  // , () => {
  //   return {
  //     useLineChartDatetime: (): ValueAndFormatted<string> => (),
  //     useLineChartPrice: (): ValueAndFormatted<string> => ({
  //       value: { value: Amounts.lg.value.toString() },
  //       formatted: { value: Amounts.lg.value.toString() },
  //     }),
  //     useLineChart: (): Partial<ReturnType<typeof useLineChart>> => ({
  //       currentIndex: { value: 1 },
  //       data: [],
  //       isActive: { value: false },
  //     }),
  //   }
  // }
)
const mockedUseLineChartPrice = charts.useLineChartPrice as jest.Mock
const mockedUseLineChart = charts.useLineChart as jest.Mock
const mockedUseLineChartDatetime = charts.useLineChartDatetime as jest.Mock

describe(PriceText, () => {
  it('renders without error', () => {
    mockedUseLineChartPrice.mockReturnValue({ value: '' })

    const tree = render(<PriceText loading={false} spotPrice={Amounts.md.value} />)

    expect(tree).toMatchSnapshot()
  })

  it('renders loading state', () => {
    mockedUseLineChartPrice.mockReturnValue({ value: '' })

    const tree = render(<PriceText loading={true} spotPrice={Amounts.md.value} />)

    expect(tree).toMatchSnapshot()
  })

  it('shows active price when scrubbing', async () => {
    mockedUseLineChartPrice.mockReturnValue({ value: { value: Amounts.sm.value.toString() } })

    const tree = render(<PriceText loading={false} spotPrice={Amounts.md.value} />)

    const text = await tree.findByTestId('price-text')
    expect(text.props.value).toBe(`$${Amounts.sm.value}.00`)
  })
})

describe(RelativeChangeText, () => {
  it('renders without error', () => {
    mockedUseLineChart.mockReturnValue({
      isActive: { value: false },
    })

    const tree = render(
      <RelativeChangeText loading={false} spotRelativeChange={Amounts.md.value} />
    )

    expect(tree).toMatchSnapshot()
  })

  it('renders loading state', () => {
    mockedUseLineChart.mockReturnValue({
      isActive: { value: false },
    })

    const tree = render(<RelativeChangeText loading={true} spotRelativeChange={Amounts.md.value} />)

    expect(tree).toMatchSnapshot()
  })

  it('shows active relative change when scrubbing', async () => {
    mockedUseLineChart.mockReturnValue({
      isActive: { value: true },
      data: [{ value: 9 }, { value: 10 }], // price goes from 9 to 10 = 10% change
      currentIndex: { value: 0 },
    })

    const tree = render(
      <RelativeChangeText loading={false} spotRelativeChange={Amounts.md.value} />
    )

    const text = await tree.findByTestId('relative-change-text')
    expect(text.props.value).toBe(`10.00%`)
  })
})

describe(DatetimeText, () => {
  it('renders without error', () => {
    mockedUseLineChartDatetime.mockReturnValue({
      value: { value: '123' },
      formatted: { value: 'Thursday, November 1st, 2023' },
    })
    const tree = render(<DatetimeText loading={false} />)

    expect(tree).toMatchSnapshot()
  })

  it('renders loading state', () => {
    mockedUseLineChartDatetime.mockReturnValue({
      value: { value: '123' },
      formatted: { value: 'Thursday, November 1st, 2023' },
    })
    const tree = render(<DatetimeText loading={true} />)

    expect(tree).toMatchSnapshot()
  })
})
