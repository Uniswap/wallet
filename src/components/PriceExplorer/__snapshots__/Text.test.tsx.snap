// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DatetimeText renders loading state 1`] = `null`;

exports[`DatetimeText renders without error 1`] = `
<TextInput
  allowFontScaling={true}
  animatedProps={
    Object {
      "text": "Thursday, November 1st, 2023",
    }
  }
  editable={false}
  maxFontSizeMultiplier={1.4}
  style={
    Array [
      Array [
        Object {
          "color": "#505A78",
          "fontFamily": "Inter-Medium",
          "fontSize": 16,
          "lineHeight": 24,
        },
      ],
    ]
  }
  underlineColorAndroid="transparent"
  value="Thursday, November 1st, 2023"
/>
`;

exports[`PriceText renders loading state 1`] = `
<View
  onLayout={[Function]}
>
  <View
    style={
      Array [
        Object {
          "alignItems": "center",
          "display": "flex",
          "flexDirection": "row",
        },
      ]
    }
  >
    <View
      style={
        Array [
          Object {
            "alignItems": "center",
            "display": "flex",
            "flexDirection": "row",
            "position": "relative",
          },
        ]
      }
    >
      <View
        accessibilityElementsHidden={true}
        importantForAccessibility="no-hide-descendants"
      >
        <Text
          allowFontScaling={true}
          maxFontSizeMultiplier={1.2}
          style={
            Array [
              Object {
                "color": "transparent",
                "display": "flex",
                "fontFamily": "Inter-SemiBold",
                "fontSize": 40,
                "lineHeight": 48,
                "opacity": 0,
              },
            ]
          }
        >
          $10,000
        </Text>
      </View>
      <View
        style={
          Array [
            Object {
              "backgroundColor": "#E8ECFB",
              "borderRadius": 4,
              "bottom": "5%",
              "display": "flex",
              "left": 0,
              "position": "absolute",
              "right": 0,
              "top": "5%",
            },
          ]
        }
      />
    </View>
  </View>
</View>
`;

exports[`PriceText renders without error 1`] = `
<TextInput
  allowFontScaling={true}
  animatedProps={
    Object {
      "text": "$55.00",
    }
  }
  editable={false}
  maxFontSizeMultiplier={1.2}
  style={
    Array [
      Array [
        Object {
          "color": "#0D111C",
          "fontFamily": "Inter-SemiBold",
          "fontSize": 40,
          "lineHeight": 48,
        },
      ],
    ]
  }
  testID="price-text"
  underlineColorAndroid="transparent"
  value="$55.00"
/>
`;

exports[`RelativeChangeText renders loading state 1`] = `
<View
  onLayout={[Function]}
>
  <View
    style={
      Array [
        Object {
          "alignItems": "center",
          "display": "flex",
          "flexDirection": "row",
        },
      ]
    }
  >
    <View
      style={
        Array [
          Object {
            "alignItems": "center",
            "display": "flex",
            "flexDirection": "row",
            "position": "relative",
          },
        ]
      }
    >
      <View
        accessibilityElementsHidden={true}
        importantForAccessibility="no-hide-descendants"
      >
        <Text
          allowFontScaling={true}
          maxFontSizeMultiplier={1.4}
          style={
            Array [
              Object {
                "color": "transparent",
                "display": "flex",
                "fontFamily": "Inter-Medium",
                "fontSize": 16,
                "lineHeight": 24,
                "opacity": 0,
              },
            ]
          }
        >
          00.00%
        </Text>
      </View>
      <View
        style={
          Array [
            Object {
              "backgroundColor": "#E8ECFB",
              "borderRadius": 4,
              "bottom": "5%",
              "display": "flex",
              "left": 0,
              "position": "absolute",
              "right": 0,
              "top": "5%",
            },
          ]
        }
      />
    </View>
  </View>
</View>
`;

exports[`RelativeChangeText renders without error 1`] = `
<View
  style={
    Array [
      Object {
        "alignItems": "stretch",
        "display": "flex",
        "flexDirection": "row",
        "justifyContent": "flex-start",
      },
    ]
  }
>
  <TextInput
    allowFontScaling={true}
    animatedProps={
      Object {
        "text": "↗",
      }
    }
    editable={false}
    maxFontSizeMultiplier={1.4}
    style={
      Array [
        Array [
          Object {
            "color": "#0D111C",
            "fontFamily": "Inter-Medium",
            "fontSize": 16,
            "lineHeight": 24,
          },
          Object {
            "color": "#40B66B",
          },
        ],
      ]
    }
    testID="relative-change-icon"
    underlineColorAndroid="transparent"
    value="↗"
  />
  <View
    style={
      Array [
        Object {
          "display": "flex",
          "flexGrow": 0,
          "flexShrink": 0,
          "height": 2,
          "width": 2,
        },
      ]
    }
  />
  <TextInput
    allowFontScaling={true}
    animatedProps={
      Object {
        "text": "55.00%",
      }
    }
    editable={false}
    maxFontSizeMultiplier={1.4}
    style={
      Array [
        Array [
          Object {
            "color": "#0D111C",
            "fontFamily": "Inter-Medium",
            "fontSize": 16,
            "lineHeight": 24,
          },
          Object {
            "color": "#40B66B",
          },
        ],
      ]
    }
    testID="relative-change-text"
    underlineColorAndroid="transparent"
    value="55.00%"
  />
</View>
`;
